sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -503,36 +503,91 @@ object DataSourceStrategy
   private def translateLeafNodeFilter(
       predicate: Expression,
       pushableColumn: PushableColumnBase): Option[Filter] = predicate match {
+
     case expressions.EqualTo(pushableColumn(name), Literal(v, t)) =>
       Some(sources.EqualTo(name, convertToScala(v, t)))
     case expressions.EqualTo(Literal(v, t), pushableColumn(name)) =>
       Some(sources.EqualTo(name, convertToScala(v, t)))
 
+    case expressions.EqualTo(Cast(e @ pushableColumn(name), _, _, _), l @ Literal(_, t))
+      if Cast.canUpCast(t, e.dataType) =>
+      val castValue = cast(l, e.dataType).eval()
+      Some(sources.EqualTo(name, convertToScala(castValue, e.dataType)))
+    case expressions.EqualTo(l @ Literal(_, t), Cast(e @ pushableColumn(name), _, _, _))
+      if Cast.canUpCast(t, e.dataType) =>
+      val castValue = cast(l, e.dataType).eval()
+      Some(sources.EqualTo(name, convertToScala(castValue, e.dataType)))
```

Comment:
Is this result correct?
```scala
spark.sql("select '1.00000' as id from range(10)").write.saveAsTable("t1")
spark.sql("select * from t1 where id = 1").show
```

Reply:
result is correct, as follow:

+---+
| id|
+---+
+---+

In fact, the above query is equal to: 
**select * from t1 where id = '1'**

Reply:
This is the result of Spark 3.3.0.
```
scala> sc.version
res2: String = 3.3.0

scala> spark.sql("select * from t1 where id = 1").show
+-------+
|     id|
+-------+
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
+-------+
```

Reply:
t1's data:

spark.sql("select '1.00000' as id from range(10)").write.saveAsTable("t1")
spark.sql("select * from t1").show

+-------+
|     id|
+-------+
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
|1.00000|
+-------+

actually spark's(and mysql) behavior:
'select * from t1 where id = 1' === 'select * from t1 where cast(id, int) = 1'

but not:
'select * from t1 where id = 1' === 'select * from t1 where id = cast(1, string)'

#### Normally, when users write the following SQL:
> select * from t1 where id = 1
#### they **usually** forget the single quote
> select * from t1 where id = '1'

It's hard to understand.

Reply:
So this is not a correct improvement.